src/block_cache.rs

```diff
@@ -0,0 +1,390 @@
+            entries: VecDeque::from(entries),
+            max_blocks_per_read: opts.max_blocks_per_read,
+            read_buf: vec![0; read_buf_len].into_boxed_slice(),
+            block_size: opts.block_size,
+            num_fs_blocks,
+        })
+    }
+
+    /// Get the number of blocks to read.
+    ///
+    /// Normally this returns `max_blocks_per_read`. If reading that
+    /// many blocks would go past the end of the filesystem, the number
+    /// is clamped to avoid that.
+    ///
+    /// # Preconditions
+    ///
+    /// `block_index` must be less than `num_fs_blocks`.
+    fn num_blocks_to_read(&self, block_index: FsBlockIndex) -> u32 {
+        assert!(block_index < self.num_fs_blocks);
+
+        // Get the index of the block right after the last block to read.
+        let end_block = block_index
+            .saturating_add(u64::from(self.max_blocks_per_read))
+            .min(self.num_fs_blocks);
+
+        // OK to unwrap: `end_block` can't be less than `block_index`.
+        let num_blocks = end_block.checked_sub(block_index).unwrap();
+
+        // OK to unwrap: the number is at most `max_blocks_per_read`,
+        // which is a `u32`.
+        u32::try_from(num_blocks).unwrap()
+    }
+
+    /// Get the cache entry for `block_size`, reading and inserting
+    /// blocks into the cache if not already present.
+    ///
+    /// If the entry is already present, it is moved to the front of the
+    /// cache to indicate it was accessed most recently.
+    ///
+    /// Otherwise, `f` is called to read a contiguous group of
+    /// blocks. Each block is inserted into the cache, with the
+    /// requested `block_index` at the front of the cache. `f` is called
+    /// only once.
+    ///
+    /// # Preconditions
+    ///
+    /// `block_index` must be less than `num_fs_blocks`.
+    pub(crate) fn get_or_insert_blocks<F>(
+        &mut self,
+        block_index: FsBlockIndex,
+        f: F,
+    ) -> Result<&[u8], Ext4Error>
+    where
+        F: FnOnce(&mut [u8]) -> Result<(), Ext4Error>,
+    {
+        assert!(block_index < self.num_fs_blocks);
+
+        // Check if the block is already cached.
+        if let Some(index) = self
+            .entries
+            .iter()
+            .position(|entry| entry.block_index == block_index)
+        {
+            // Move the entry to the front of the cache if it's not
+            // already there.
+            if index != 0 {
+                let entry = self.entries.remove(index).unwrap();
+                self.entries.push_front(entry);
+            }
+
+            // Return the cached block data.
+            return Ok(&*self.entries[0].data);
+        }
+
+        let block_size = self.block_size.to_usize();
+
+        // Get the number of blocks/bytes to read.
+        let num_blocks = self.num_blocks_to_read(block_index);
+        let num_bytes = usize_from_u32(num_blocks)
+            .checked_mul(block_size)
+            .unwrap_or(block_size);
```

Comment:
Why `unwrap_or(block_size)`? Probably wants a comment, because checking that this makes sense is taking me a bit...

* If the multiplication overflows, we'll have num_blocks (almost definitely) > 1, and num_bytes == one block's worth.
* L194 we'll read one blocks worth of bytes into read_buf... I'm actually unclear what happens to the rest of the buffer: I don't see anything that indicates that it's zero'd
* when we iterate on L199, we'll `insert_block` for an index outside what we actually read.  Not clear to me what happens.

Reply:
Changed it to return an error (added as a new commit to help with review)